File: backend_spring/src/main/java/org/kuroneko/swn_char_gen/backend_spring/services/CharacterService.java
/*
 * SWN Char Gen © 2025. by Tod Casasent is licensed under CC BY-NC-SA 4.0. To view a copy of this license, visit https://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * Rules and values associated with the SWN system are the copyright of Kevin Crawford, Sine Nominee Publishing.
 * https://sine-nomine-publishing.myshopify.com/
 * Used in accordance with his Discord message of limiting content to the free version of the rules.
 * Specifically, the SWNSystem.java and swn_system.py game strings, values, and relationship are the copyright of Kevin Crawford, Sine Nominee Publishing.
 */

package org.kuroneko.swn_char_gen.backend_spring.services;

import org.kuroneko.swn_char_gen.backend_spring.models.AttributeEnum;
import org.kuroneko.swn_char_gen.backend_spring.models.CharacterStore;
import org.kuroneko.swn_char_gen.backend_spring.models.DetailEnum;
import org.kuroneko.swn_char_gen.backend_spring.models.SWNChar;
import org.springframework.stereotype.Service;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import jakarta.servlet.http.HttpSession;

import java.io.IOException;
import java.util.Map;
import java.util.UUID;

/**
 * Service for SWN character operations.
 */
@Service
public class CharacterService
{

	/**
	 * Get the current HTTP session.
	 *
	 * @return The current HTTP session
	 */
	private HttpSession getCurrentSession()
	{
		ServletRequestAttributes attr = (ServletRequestAttributes) RequestContextHolder.currentRequestAttributes();
		return attr.getRequest().getSession(true);
	}

	/**
	 * Get or create a session ID.
	 *
	 * @return The session ID
	 */
	public String getOrCreateSessionId()
	{
		HttpSession session = getCurrentSession();
		String sessionId = (String) session.getAttribute("session_id");
		if (sessionId == null)
		{
			sessionId = UUID.randomUUID().toString();
			session.setAttribute("session_id", sessionId);
		}
		return sessionId;
	}

	/**
	 * Upload a character from JSON data.
	 *
	 * @param characterData A Map containing character data
	 * @return The uploaded character data as a Map
	 */
	public Map<String, Object> uploadCharacter(Map<String, Object> characterData) throws IOException
	{
		String sessionId = getOrCreateSessionId();
		// Create a new character from the uploaded data
		SWNChar character = new SWNChar(characterData);
		// Store the character
		CharacterStore.getInstance().storeCharacter(sessionId, character);
		return character.toMap();
	}

	/**
	 * Create a new character.
	 *
	 * @return The new character data as a Map
	 */
	public Map<String, Object> createNewCharacter()
	{
		String sessionId = getOrCreateSessionId();
		SWNChar character = new SWNChar();
		CharacterStore.getInstance().storeCharacter(sessionId, character);
		return character.toMap();
	}

	/**
	 * Get the current character.
	 *
	 * @return The current character data as a Map
	 */
	public Map<String, Object> getCharacter()
	{
		String sessionId = getOrCreateSessionId();
		CharacterStore store = CharacterStore.getInstance();

		if (store.characterMissing(sessionId))
		{
			SWNChar character = new SWNChar();
			store.storeCharacter(sessionId, character);
			return character.toMap();
		}
		else
		{
			SWNChar character = store.getCharacter(sessionId);
			return character.toMap();
		}
	}

	/**
	 * Roll attributes for the current character.
	 *
	 * @return The updated character data as a Map
	 */
	public Map<String, Object> rollAttributes()
	{
		String sessionId = getOrCreateSessionId();
		CharacterStore store = CharacterStore.getInstance();

		if (store.characterMissing(sessionId))
		{
			SWNChar character = new SWNChar();
			store.storeCharacter(sessionId, character);
			character.rollAttributes();
			return character.toMap();
		}
		else
		{
			SWNChar character = store.getCharacter(sessionId);
			character.rollAttributes();
			return character.toMap();
		}
	}

	/**
	 * Change an attribute for the current character.
	 *
	 * @param attributeName The name of the attribute to change
	 * @return The updated character data as a Map
	 * @throws IllegalArgumentException if no character is found or the attribute is invalid
	 */
	public Map<String, Object> changeAttribute(String attributeName)
	{
		String sessionId = getOrCreateSessionId();
		CharacterStore store = CharacterStore.getInstance();

		if (store.characterMissing(sessionId))
		{
			throw new IllegalArgumentException("No character found");
		}

		SWNChar character = store.getCharacter(sessionId);
		AttributeEnum attributeEnum = AttributeEnum.valueOf(attributeName.toUpperCase());
		character.changeOneAttribute(attributeEnum);
		return character.toMap();
	}

	/**
	 * Set a detail for the current character.
	 *
	 * @param detailName  The name of the detail to set
	 * @param detailValue The value to set
	 * @return The updated character data as a Map
	 * @throws IllegalArgumentException if no character is found or the detail is invalid
	 */
	public Map<String, Object> setDetail(String detailName, String detailValue)
	{
		String sessionId = getOrCreateSessionId();
		CharacterStore store = CharacterStore.getInstance();

		if (store.characterMissing(sessionId))
		{
			throw new IllegalArgumentException("No character found");
		}

		SWNChar character = store.getCharacter(sessionId);
		DetailEnum detailEnum = DetailEnum.valueOf(detailName.toUpperCase());
		character.setDetail(detailEnum, detailValue);
		return character.toMap();
	}
}

File: backend_flask/application/services/character_service.py
#  SWN Char Gen © 2025. by Tod Casasent is licensed under CC BY-NC-SA 4.0. To view a copy of this license, visit https://creativecommons.org/licenses/by-nc-sa/4.0/
# 
#  Rules and values associated with the SWN system are the copyright of Kevin Crawford, Sine Nominee Publishing.
#  https://sine-nomine-publishing.myshopify.com/
#  Used in accordance with his Discord message of limiting content to the free version of the rules.
#  Specifically, the SWNSystem.java and swn_system.py game strings, values, and relationship are the copyright of Kevin Crawford, Sine Nominee Publishing.

"""
Character service module for the Stars Without Number Character Generator.

This module provides services for creating, retrieving, and manipulating
character data. It acts as an intermediary between the API routes and the
character store, handling the business logic for character operations.
"""

from typing import Dict, Any
from flask import session
import uuid
from swn_char.swn_char import SWNChar
from swn_char.swn_system import AttributeEnum, DetailEnum
from application.models.character_store import CharacterStore

class CharacterService:
    """
    Service class for character operations.
    
    This class provides static methods for creating, retrieving, and manipulating
    character data. It uses the CharacterStore to persist character data between
    requests.
    """
    @staticmethod
    def get_or_create_session_id() -> str:
        """
        Get the current session ID or create a new one if it doesn't exist.
        
        This method ensures that each user has a unique session ID that can be
        used to associate them with their character data.
        
        Returns:
            str: The session ID.
        """
        if 'session_id' not in session:
            session['session_id'] = str(uuid.uuid4())
        return session['session_id']
        
    @staticmethod
    def upload_character(character_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Upload a character from JSON data.
        
        Args:
            character_data: A dictionary containing character data
            
        Returns:
            The uploaded character data as a dictionary
        """
        session_id = CharacterService.get_or_create_session_id()
        # Create a new character from the uploaded data
        character = SWNChar(character_data)
        # Store the character
        CharacterStore.get_instance().store_character(session_id, character)
        return character.to_dict()

    @staticmethod
    def create_new_character() -> Dict[str, Any]:
        """
        Create a new character with default values.
        
        This method creates a new character with default values and stores it
        in the character store associated with the current session.
        
        Returns:
            Dict[str, Any]: The newly created character data as a dictionary.
        """
        session_id = CharacterService.get_or_create_session_id()
        character = SWNChar()
        CharacterStore.get_instance().store_character(session_id, character)
        return character.to_dict()

    @staticmethod
    def get_character() -> Dict[str, Any]:
        """
        Get the character associated with the current session.
        
        This method retrieves the character associated with the current session.
        If no character exists, a new one is created.
        
        Returns:
            Dict[str, Any]: The character data as a dictionary.
        """
        session_id = CharacterService.get_or_create_session_id()
        store = CharacterStore.get_instance()
        
        if not store.has_character(session_id):
            character = SWNChar()
            store.store_character(session_id, character)
        else:
            character = store.get_character(session_id)
            
        return character.to_dict()

    @staticmethod
    def roll_attributes() -> Dict[str, Any]:
        """
        Roll attributes for the character associated with the current session.
        
        This method generates random values for all attributes of the character
        associated with the current session. If no character exists, a new one
        is created.
        
        Returns:
            Dict[str, Any]: The updated character data as a dictionary.
        """
        session_id = CharacterService.get_or_create_session_id()
        store = CharacterStore.get_instance()
        
        if not store.has_character(session_id):
            character = SWNChar()
            store.store_character(session_id, character)
        else:
            character = store.get_character(session_id)
            
        character.roll_attributes()
        return character.to_dict()

    @staticmethod
    def change_attribute(attribute_name: str) -> Dict[str, Any]:
        """
        Change a specific attribute of the character associated with the current session.
        
        This method changes the value of a specific attribute of the character
        associated with the current session.
        
        Args:
            attribute_name: The name of the attribute to change.
            
        Returns:
            Dict[str, Any]: The updated character data as a dictionary.
            
        Raises:
            ValueError: If no character is found for the current session.
        """
        session_id = session.get('session_id')
        store = CharacterStore.get_instance()
        
        if not session_id or not store.has_character(session_id):
            raise ValueError("No character found")
            
        character = store.get_character(session_id)
        attribute_enum = getattr(AttributeEnum, attribute_name.upper())
        character.change_one_attribute(attribute_enum)
        return character.to_dict()

    @staticmethod
    def set_detail(detail_name: str, detail_value: str) -> Dict[str, Any]:
        """
        Set a detail of the character associated with the current session.
        
        This method sets a specific detail (like name) of the character
        associated with the current session.
        
        Args:
            detail_name: The name of the detail to set.
            detail_value: The value to set for the detail.
            
        Returns:
            Dict[str, Any]: The updated character data as a dictionary.
            
        Raises:
            ValueError: If no character is found for the current session.
        """
        session_id = session.get('session_id')
        store = CharacterStore.get_instance()
        
        if not session_id or not store.has_character(session_id):
            raise ValueError("No character found")
            
        character = store.get_character(session_id)
        detail_enum = getattr(DetailEnum, detail_name.upper())
        character.set_detail(detail_enum, detail_value)
        return character.to_dict()
